File: Frontend/src/lib/services/strategies.ts
import { CacheStrategies, CacheStrategy, Strategy } from "@/interfaces/entities/strategy";
import { Result } from "@/types/result";
import { invoke } from "@tauri-apps/api/core";

// db
async function insert_strategy(
    env: string,
    strategy: Strategy
): Promise<Result<Strategy>> {
    try {
        strategy.id = 0;

        const data = await invoke<Strategy>("insert_strategy", {
            env,
            strategy,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function select_strategy(
    env: string,
    strategy?: Strategy
): Promise<Result<Strategy>> {
    try {
        const data = await invoke<Strategy>("select_strategy", {
            env,
            strategy,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function select_strategies(env: string): Promise<Result<Strategy[]>> {
    try {
        let data = await invoke<Strategy[]>("select_strategies", {
            env,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function update_strategy(
    env: string,
    strategy: Strategy
): Promise<Result<Strategy>> {
    try {
        const data = await invoke<Strategy>("update_strategy", {
            env,
            strategy,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function delete_strategy(
    env: string,
    strategy: Strategy
): Promise<Result<number>> {
    try {
        const data = await invoke<number>("delete_strategy", {
            env,
            strategy,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

// cache
async function get_active_strategy(
    env: string,
    strategy?: Strategy
): Promise<Result<CacheStrategy | null>> {
    try {
        const data = await invoke<CacheStrategy | null>("get_active_strategy", {
            env,
            strategy,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function get_active_strategies(
    env: string,
    strategy?: Strategy
): Promise<Result<CacheStrategies>> {
    try {
        const data = await invoke<CacheStrategies>(
            "get_active_strategies",
            { env, strategy }
        );

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function start_active_strategies(env: string): Promise<Result<void>> {
    try {
        await invoke("start_active_strategies", {
            env,
        });

        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

async function stop_active_strategies(env: string): Promise<Result<void>> {
    try {
        await invoke("stop_active_strategies", {
            env,
        });

        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

export const strategyService = {
    // db
    insert: (env: string, strategy: Strategy): Promise<Result<Strategy>> =>
        insert_strategy(env, strategy),

    select: (env: string, strategy?: Strategy): Promise<Result<Strategy>> =>
        select_strategy(env, strategy),

    selectAll: (env: string): Promise<Result<Strategy[]>> =>
        select_strategies(env),

    update: (env: string, strategy: Strategy): Promise<Result<Strategy>> =>
        update_strategy(env, strategy),

    delete: (env: string, strategy: Strategy): Promise<Result<number>> =>
        delete_strategy(env, strategy),

    // cache
    getActive: (env: string, strategy?: Strategy): Promise<Result<CacheStrategy | null>> =>
        get_active_strategy(env, strategy),

    getActiveAll: (
        env: string,
        strategy?: Strategy
    ): Promise<Result<CacheStrategies>> =>
        get_active_strategies(env, strategy),

    startActive: (env: string): Promise<Result<void>> =>
        start_active_strategies(env),

    stopActive: (env: string): Promise<Result<void>> =>
        stop_active_strategies(env),
};

File: Frontend/src-tauri/src/command/strategies.rs
use crate::{
    core::{
        delete_strategy_core, get_active_strategies_core, get_active_strategy_core,
        insert_strategy_core, select_strategies_core, select_strategy_core,
        start_active_strategies_core, stop_active_strategies_core, update_strategy_core,
    },
    models::{
        entities::strategies::Model,
        structs::{CacheStrategies, StrategyRequest},
    },
};

// db
#[tauri::command]
pub async fn insert_strategy(env: String, strategy: StrategyRequest) -> Result<Model, String> {
    insert_strategy_core(env, strategy).await
}

#[tauri::command]
pub async fn select_strategy(
    env: String,
    strategy: Option<StrategyRequest>,
) -> Result<Model, String> {
    select_strategy_core(env, strategy).await
}

#[tauri::command]
pub async fn select_strategies(
    env: String,
    strategy: Option<StrategyRequest>,
) -> Result<Vec<Model>, String> {
    select_strategies_core(env, strategy).await
}

#[tauri::command]
pub async fn update_strategy(env: String, strategy: StrategyRequest) -> Result<Model, String> {
    update_strategy_core(env, strategy).await
}

#[tauri::command]
pub async fn delete_strategy(env: String, strategy: StrategyRequest) -> Result<u64, String> {
    delete_strategy_core(env, strategy).await
}

// cache
#[tauri::command]
pub async fn get_active_strategy(
    env: String,
    strategy: Option<StrategyRequest>,
) -> Result<Model, String> {
    get_active_strategy_core(env, strategy).await
}

#[tauri::command]
pub async fn get_active_strategies(
    env: String,
    strategy: Option<StrategyRequest>,
) -> Result<Option<CacheStrategies>, String> {
    get_active_strategies_core(env, strategy).await
}

#[tauri::command]
pub async fn start_active_strategies(env: String) -> Result<Vec<Model>, String> {
    start_active_strategies_core(env).await
}

#[tauri::command]
pub async fn stop_active_strategies(env: String) -> Result<(), String> {
    stop_active_strategies_core(env).await
}
